dask_match/repartition.py

```diff
@@ -1,22 +1,101 @@
 
     @property
     def _meta(self):
         return self.frame._meta
 
+    def _divisions(self):
+        if self.n is not None:
+            return self.simplify()._divisions()
+        return self.new_divisions
+
+    def _simplify_down(self):
+        if type(self) != Repartition:
+            # This simplify logic should not be inherited
+            return None
+        if self.n is not None:
+            if self.n < self.frame.npartitions:
+                return RepartitionToFewer(self.frame, self.n)
+            else:
+                original_divisions = divisions = pd.Series(
+                    self.frame.divisions
+                ).drop_duplicates()
+                if self.frame.known_divisions and (
+                    is_datetime64_any_dtype(divisions.dtype)
+                    or is_numeric_dtype(divisions.dtype)
+                ):
+                    npartitions = self.n
+                    df = self.frame
+                    if is_datetime64_any_dtype(divisions.dtype):
+                        divisions = divisions.values.astype("float64")
+
+                    if is_series_like(divisions):
+                        divisions = divisions.values
+
+                    n = len(divisions)
+                    divisions = np.interp(
+                        x=np.linspace(0, n, npartitions + 1),
+                        xp=np.linspace(0, n, n),
+                        fp=divisions,
+                    )
+                    if is_datetime64_any_dtype(original_divisions.dtype):
+                        divisions = methods.tolist(
+                            pd.Series(divisions).astype(original_divisions.dtype)
+                        )
+                    elif np.issubdtype(original_divisions.dtype, np.integer):
+                        divisions = divisions.astype(original_divisions.dtype)
+
+                    if isinstance(divisions, np.ndarray):
+                        divisions = divisions.tolist()
+
+                    divisions = list(divisions)
+                    divisions[0] = df.divisions[0]
+                    divisions[-1] = df.divisions[-1]
+
+                    # Ensure the computed divisions are unique
+                    divisions = list(unique(divisions[:-1])) + [divisions[-1]]
+                    return RepartitionDivisions(df, divisions, self.force)
+                else:
+                    return RepartitionToMore(self.frame, self.n)
+        elif self.new_divisions:
+            return RepartitionDivisions(self.frame, self.new_divisions, self.force)
+        else:
+            raise NotImplementedError()
+
+    def _simplify_up(self, parent):
+        # Move filter or column projection before repartitioning
+        if isinstance(parent, (Filter, Projection)) and not isinstance(
+            parent.operands[1], Expr
+        ):
+            op = type(parent)(self.frame, *parent.operands[1:])
+            return type(self)(op, *self.operands[1:])
```

Comment:
```python
if isinstance(parent, Projection):
    return self.frame[parent.columns].repartition(*self.operands[1:])
if isinstance(parent, Filter):
    ...
```

Reply:
Or I guess we want to use the proper Op, given that we may already have lowered.  In practice though I expect `simplify_up` to be called before `simplify_down`.  It's best not to rely on that though.